core/domain/feature_flag_services.py

```diff
@@ -0,0 +1,252 @@
+                    'feature_stage': feature_with_init_settings.feature_stage,
+                    'force_enable_for_all_users': (
+                        feature_model.force_enable_for_all_users),
+                    'rollout_percentage': feature_model.rollout_percentage,
+                    'user_group_ids': feature_model.user_group_ids,
+                    'last_updated': last_updated
+                })
+            )
+
+    for feature_name in feature_names_list:
+        if feature_name not in feature_name_to_feature_flag_model_dict:
+            feature_name_to_feature_flag_model_dict[feature_name] = None
+
+    return feature_name_to_feature_flag_model_dict
+
+
+def is_feature_flag_enabled(
+    user_id: Optional[str],
+    feature_name: str,
+    feature_flag: Optional[feature_flag_domain.FeatureFlag] = None
+) -> bool:
+    """Returns True if feature is enabled for the given user else False.
+
+    Args:
+        user_id: str|None. The id of the user, if logged-out user then None.
+        feature_name: str. The name of the feature flag that needs to
+            be evaluated.
+        feature_flag: FeatureFlag. The feature flag domain model.
+
+    Returns:
+        bool. True if the feature is enabled for the given user else False.
+    """
+    if feature_flag is None:
+        feature_flag = registry.Registry.get_feature_flag(feature_name)
+    current_server = feature_flag_domain.get_server_mode()
+
+    if (
+        current_server == feature_flag_domain.ServerMode.TEST and
+        feature_flag.feature_stage == feature_flag_domain.ServerMode.DEV.value
+    ):
+        return False
+
+    if (
+        current_server == feature_flag_domain.ServerMode.PROD and
+        feature_flag.feature_stage in (
+            feature_flag_domain.ServerMode.DEV.value,
+            feature_flag_domain.ServerMode.TEST.value)
+    ):
+        return False
+
+    if feature_flag.force_enable_for_all_users:
+        return True
+    if user_id:
+        random_bytes = secrets.token_bytes(16)
+        salt = feature_name.encode('utf-8') + random_bytes
+        hashed_user_id = hashlib.sha256(
+            user_id.encode('utf-8') + salt).hexdigest()
+        hash_value = int(hashed_user_id, 16)
+        mod_result = hash_value % 1000
+        threshold = (feature_flag.rollout_percentage / 100) * 1000
+        return bool(mod_result < threshold)
+    return False
+
+
+def evaluate_all_feature_flag_values(user_id: Optional[str]) -> Dict[str, bool]:
+    """Evaluates and returns the value of feature flags.
+
+    Args:
+        user_id: str|None. The id of the user, if logged-out user then None.
+
+    Returns:
+        dict. The keys are the feature names and the values are boolean
+        results of corresponding flags.
+    """
+    result_dict = {}
+    feature_flag_dicts = get_all_feature_flag_dicts()
+    for feature_dict in feature_flag_dicts:
+        feature_domain_model = feature_flag_domain.FeatureFlag.from_dict(
```

Comment:
This name seems wrong, it's a domain object not a model right?

Reply:
Updated